### PR TITLE
esp32s3/wifi: add support to softAP (softAP and softAP + STA mode)

### PR DESCRIPTION
**Pull request by tmedicci**
## Summary

Add support to softAP (softAP and softAP + STA) mode on ESP32-S3.

## Impact

Besides implementing softAP support, this MR also fixes the bug reported at https://github.com/apache/incubator-nuttx/issues/7193 and https://github.com/apache/nuttx/issues/7857 for the ESP32-S3 (needs backport the same fix to other SoCs [WIP]).

## Testing

CI testing and `sta_softap` defconfig + `iperf` long-term testing.

### CODE REVIEW

**Review by lucasssvaz — APPROVED** [review]

**Review by acassis — APPROVED** [review]

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c [hunk 1]
```diff
@@ -4642,7 +4713,18 @@ int esp_wifi_sta_start(void)
...
+    {
+      mode = WIFI_MODE_APSTA;
+    }
+  else
+    {
+      mode = WIFI_MODE_STA;
+    }
+#else
   mode = WIFI_MODE_STA;
+#endif /* ESP32S3_WLAN_HAS_SOFTAP */
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c:4727]:
```suggestion
#ifdef ESP32S3_WLAN_HAS_SOFTAP
  if (g_softap_started)
    {
      mode = WIFI_MODE_APSTA;
    }
  else
#endif /* ESP32S3_WLAN_HAS_SOFTAP */
    {
      mode = WIFI_MODE_STA;
    }
```

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c [hunk 2]
```diff
@@ -4701,10 +4783,37 @@ int esp_wifi_sta_stop(void)
...
   wlinfo("OK to stop Wi-Fi station\n");
 
   esp_wifi_lock(false);
   return OK;
+
+#ifdef ESP32S3_WLAN_HAS_SOFTAP
+errout_set_mode:
+  esp_wifi_lock(false);
+  return ret;
+#endif /* ESP32S3_WLAN_HAS_SOFTAP */
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c:4816]:
```suggestion
  wlinfo("OK to stop Wi-Fi station\n");

errout:
  esp_wifi_lock(false);
  return ret;
```
and use `errout` label

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c [hunk 3]
```diff
@@ -5645,6 +5923,802 @@ int esp_wifi_sta_rssi(struct iwreq *iwr, bool set)
...
+    {
+      mode = WIFI_MODE_APSTA;
+    }
+  else
+    {
+      mode = WIFI_MODE_AP;
+    }
+#else
+  mode = WIFI_MODE_AP;
+#endif /* ESP32S3_WLAN_HAS_STA */
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c:5971]:
```suggestion
#ifdef ESP32S3_WLAN_HAS_STA
  if (g_sta_started)
    {
      mode = WIFI_MODE_APSTA;
    }
  else
#endif /* ESP32S3_WLAN_HAS_STA */
    {
      mode = WIFI_MODE_AP;
    }
```

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c [hunk 4]
```diff
@@ -5645,6 +5923,802 @@ int esp_wifi_sta_rssi(struct iwreq *iwr, bool set)
...
+  g_softap_started = true;
+
+  wlinfo("OK to start Wi-Fi SoftAP\n");
+
+  esp_wifi_lock(false);
+  return OK;
+
+errout_set_mode:
+  esp_wifi_lock(false);
+  return ret;
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c:5998]: ditto

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c [hunk 5]
```diff
@@ -5645,6 +5923,802 @@ int esp_wifi_sta_rssi(struct iwreq *iwr, bool set)
...
+  wlinfo("OK to stop Wi-Fi SoftAP\n");
+
+  esp_wifi_lock(false);
+  return OK;
+
+#ifdef ESP32S3_WLAN_HAS_STA
+errout_set_mode:
+  esp_wifi_lock(false);
+  return ret;
+#endif /* ESP32S3_WLAN_HAS_STA */
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c:6060]: ditto

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c (hunk 3 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c]: done ;)

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c (hunk 2 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c]: Partially done:

If the label `errout`  is present outside the `#ifdef`, the compiler will throw a warning:

```
chip/esp32s3_wifi_adapter.c:4804:1: error: label 'errout' defined but not used [-Werror=unused-label]
 4804 | errout:
```

I changed the label itself and removed the redundant `esp_wifi_lock`. Please take a look if it fits ;)

```
  wlinfo("OK to stop Wi-Fi station\n");

#ifdef ESP32S3_WLAN_HAS_SOFTAP
errout:
#endif /* ESP32S3_WLAN_HAS_SOFTAP */

  esp_wifi_lock(false);
  return ret;
 ```

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c (hunk 4 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c]: done ;)

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c (hunk 5 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.c]: Please check https://github.com/apache/nuttx/pull/8818#discussion_r1138559802

**Review by pkarashchenko — APPROVED** [review]

File: arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.h [hunk 6]
```diff
@@ -476,6 +486,344 @@ int esp_wifi_sta_country(struct iwreq *iwr, bool set);
...
+ *
+ * Input Parameters:
+ *   mac  - MAC address buffer pointer
+ *
+ * Returned Value:
+ *   0 if success or -1 if fail
+ *
+ ****************************************************************************/
+
+int esp_wifi_softap_read_mac(uint8_t *mac);
```
[the remark this answers is not in the document]
Comment (tmedicci) [reply arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.h:592]: I'm preparing a PR with minor changes and fixes considering all Espressif's SoCs and I'll address this change in there, alright?

[the remark this answers is not in the document]
Comment (pkarashchenko) [reply arch/xtensa/src/esp32s3/esp32s3_wifi_adapter.h:592]: Sure

**Pull request merged**